### DISCUSSION TITLE
Rewriting another Next.js collides with _next/static assets and causes 404

### DISCUSSION THREAD

**Discussion by bdcorps** [vercel/next.js#43212] [Help]
I have a Next.js app (App1) and I want to be able to rewrite on `/*` to another Next.js app hosted here (App2): https://chakra-ui-blog.vercel.app. I do not have control over code of the App2. 

In my App1, I set up a middleware:
```
import { NextResponse } from "next/server";

import type { NextRequest } from "next/server";

export const config = {
  matcher: [
    "/(.*)",
  ],
};

export async function middleware(req: NextRequest) {
  const url = req.nextUrl.clone();
  const { pathname } = req.nextUrl;

  console.log("asking for: ", pathname);

  if (pathname.includes("_app-7bdae2848c07e446.js")) {
    console.log("yes i see this file");
  }

  return NextResponse.rewrite(
    new URL(pathname, `https://chakra-ui-blog.vercel.app`)
  );
}
```

Upon running the app, you can see that some of the files are returning 404 after the rewrite.

<img width="896" alt="Screen Shot 2022-11-21 at 10 06 47 PM" src="https://user-images.githubusercontent.com/6132555/203211433-c9ce2155-cc92-47a6-a291-7a8dc4327452.png">

As noted in the code sample, `_app-7bdae2848c07e446.js` file is never seen by the middleware and simply hits 404.

I have already looked at the Next.js Multi Zones example and this is different from that. 

Is there a way to fix this?

**Comment by icyJoseph** [comment]
What next.js version are you using? I did:

```ts
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  console.log("Middleware", { pathname });

  const rewrite = new URL(pathname, "http://localhost:3001");

  return NextResponse.rewrite(rewrite);
}

```

And now I see the app at 3001 on my app on 3000, zero problems. Tested both in dev and production modes.

## Edit

`_app` requests and JS chunks for a page are not loaded with the above. It does however work with `appDir` 👍 , for now

**Comment by bdcorps** [comment]
@icyJoseph Next.js version - `next@13.0.5-canary.4`

Tried your code sample but ran into the same problem as described above. The HTML content loads but the CSS and JS do not.

If you run this, you should be able to notice the 404 missing files in the `Network` tab.

``` js
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  console.log("Middleware", { pathname });

  const rewrite = new URL(pathname, "https://chakra-ui-info-product.vercel.app"); // or https://chakra-ui-blog.vercel.app

  return NextResponse.rewrite(rewrite);
}
```

**Reply by icyJoseph** [reply]
Aha I see the next app I had in 3001, was using the appDir, and had no `_app` present at all... Mmm it is a bit unfortunate that those `_next` internals are not passed to the middleware though, nor through `next.config.js`, I guess that's where the limitation is. 

I have however managed to use a custom server to mutate incoming paths 🙈😰 really bad...

https://user-images.githubusercontent.com/21013447/203332190-707d249a-8a20-416b-a110-6f52aac9387a.mov

Since I don't like this solution I won't post the code, but you can start from https://github.com/vercel/next.js/tree/canary/examples/custom-server and there you could prefix all requests with some label, remove that label in the middleware and rewrite. 

Do remember that using a custom server is discouraged and it prevents you from hosting on Vercel, IIRC.

**Reply by bdcorps** [reply]
Yeah, it really limits the usage of middleware if some files aren't going through it.

I'll open up a bug report regarding this. Thank you for looking through the problem :)

**Reply by icyJoseph** [reply]
Not sure it is a bug though, I think if we read the middleware feature set, we see that the requests that arrive to middleware are per resource, _app and the JS bundles that drive pages are in a grey area at best. This should be a feature request instead.
